executor/README.md

```diff
@@ -0,0 +1,226 @@
+| 204  | No Content          | The request was succesfully processed. |
+| 400  | Malformed           | The request body contains an unknown key, or a key's value is the wrong type. |
+| 503  | Service Unavailable | The request was valid, and received, but there was an (ostensibly temporary) failure with mesos itself. |
+
+### Example use cases
+
+Below are some examples of how to use the above features to solve some common problems.
+
+#### Conditionally preventing command execution
+
+In cases where only one copy of a certain job should be running, before and after commands could be used to acquire and release a lock. Assuming you've written some scripts for managing your lock, you could execute them by passing the additional data to the cook scheduler's job creation API:
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./acquire-zk-lock.sh my-service"
+		"guard": true
+	}],
+	"after_commands":[{
+		"value": "./release-zk-lock.sh my-service"
+	}]
+}
+```
+
+The `guard` flag is used to prevent execution of later commands if we can't acquire the lock (in this case, `acquire-zk-lock.sh` script will return a non-zero exit code). Note that the scripts for manging the lock could be completely reusable, and that the primary command did not need to change.
+
+#### Monitoring a running command
+
+Perhaps you'd like to tail a log created by your primary command. You could use an asynchronous before command to so. It will be started before your primary command, and continue running in the background until the primary command (and all other synchronous commands) finish.
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./my-monitoring-command",
+		"async": true
+	}]
+}
+```
+
+#### Dynamically setting environment variables
+
+In order to dynamically set environment variables for later commands, you can use the executor HTTP API from a before command.
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./set-env-vars.sh"
+	}]	
+}
+```
+
+The before command would need to executor something like the following:
+
+```bash
+curl -XPATCH \
+	-d '{"env": {"MY_ENV_VAR": "/hello/world", "MY_OLD_VAR": null}}' \
+	127.0.0.1:$EXECUTOR_PORT0/task/$TASK_ID
+```
+
+This would set `MY_ENV_VAR` for your primary command (and all commands after `set-env-vars.sh`. 
+
+#### Update progress and custom failure messages
+
+In order to update progress or add a custom failure message, you can use the executor HTTP API from any command, with something like the following:
+
+```bash
+curl -XPATCH \
+	-d '{"progress": 100, "message": "a custom failure message"}' \
+	127.0.0.1:$EXECUTOR_PORT0/task/$TASK_ID
+```
+
+#### Sending an email on completion
+
+In order to send an email once your primary command has finished, you can use an after command:
+
+```json
+{
+	"after_commands": [{
+		"value": "mail -s 'your job failed, sorry' bob@hello.com"
```

Comment:
Do the after commands have access to exit codes of commands run before? i.e. if this script actually wanted to notify the user on failure, how would it know the command failed?